File: 121_best_time_stock.py
from typing import List


def maxProfit(prices: List[int]) -> int:
    """
    Say you have an array for which the ith element is the price of a given
    stock on day i. If you were only permitted to complete at most one
    transaction (i.e., buy one and sell one share of the stock), design an
    algorithm to find the maximum profit.
    Note that you cannot sell a stock before you buy one.
    """
    min_price = float("inf")
    max_profit = 0
    for i in prices:
        if i < min_price:
            min_price = i
        elif i - min_price > max_profit:
            max_profit = i - min_price
    return max_profit


arr = [7, 1, 5, 3, 6, 4]  # 5
print(maxProfit(arr))

arr2 = [7, 5, 4, 3, 2, 1]  # 0
print(maxProfit(arr2))

arr3 = [2, 4, 1, 2]  # 2
print(maxProfit(arr3))

File: 121_best_time_stock.js
/**
 * Say you have an array for which the ith element is the price of a given
 * stock on day i. If you were only permitted to complete at most one
 * transaction (i.e., buy one and sell one share of the stock), design an
 * algorithm to find the maximum profit.
 * Note that you cannot sell a stock before you buy one.
 * @param   {number[]} prices
 * @return  {number}
 */
const maxProfit = prices => {
  let minPrice = Number.MAX_SAFE_INTEGER
  return prices.reduce((maxProfit, e) => {
    if (e < minPrice) minPrice = e
    else if (e - minPrice > maxProfit) maxProfit = e - minPrice
    return maxProfit
  }, 0)
}

const arr = [7, 1, 5, 3, 6, 4] // 5
console.log(maxProfit(arr))

const arr2 = [7, 5, 4, 3, 2, 1] // 0
console.log(maxProfit(arr2))

const arr3 = [2, 4, 1, 2] // 2
console.log(maxProfit(arr3))
